src/lib/jobrunner.rs

```diff
@@ -0,0 +1,409 @@
+use std::{
+    error::Error,
+    fs::{self, remove_file, File},
+    io::{Read, Seek, SeekFrom, Write},
+    os::unix::io::{AsRawFd, RawFd},
+    path::PathBuf,
+    process::{self, Child, Command},
+    sync::Arc,
+    thread,
+};
+
+use lettre::{sendmail::SendmailTransport, EmailAddress, Envelope, SendableEmail, Transport};
+use nix::{
+    fcntl::{fcntl, FcntlArg, OFlag},
+    poll::{poll, PollFd, PollFlags},
+};
+use tempfile::{tempdir, tempfile, NamedTempFile, TempDir};
+use whoami::{hostname, username};
+
+use crate::{queue::QueueJob, Snare};
+
+const READBUF: usize = 8 * 1024; // Bytes. Should be >= PIPE_BUF for performance reasons.
+const WAIT_TIMEOUT: i32 = 1; // Seconds.
+const RECHECK_TIMEOUT: i32 = 5; // Seconds.
+const DEFAULT_TIMEOUT: i32 = 60; // Seconds. How often to check the queue "just in case?
+
+struct JobRunner {
+    snare: Arc<Snare>,
+    /// The running jobs, with `0..num_running` entries.
+    running: Vec<Option<Job>>,
+    /// How many `Some` entries are there in `self.running`?
+    num_running: usize,
+    /// The running jobs, with `0..2 *num_running + 1` entries. Each pair of entries are (stderr,
+    /// stdout) for the corresponding `running` `Job` (i.e. `running[0]` has its stderr entry at
+    /// `pollfds[0]` and stdout entry at `pollfds[1]`). The `+ 1` entry is the event file
+    /// descriptor that allows the HTTP server thread to wake up the JobRunner thread.
+    pollfds: Vec<PollFd>,
+}
+
+impl JobRunner {
+    fn new(snare: Arc<Snare>) -> Result<Self, Box<dyn Error>> {
+        if snare.config.maxjobs == std::usize::MAX {
+            eprintln!(
+                "The maximum number of possible jobs is {}",
+                std::usize::MAX - 1
+            );
+            process::exit(1);
+        }
+        let mut running = Vec::with_capacity(snare.config.maxjobs);
+        running.resize_with(snare.config.maxjobs, || None);
+        let mut pollfds = Vec::with_capacity(snare.config.maxjobs * 2 + 1);
+        pollfds.resize_with(snare.config.maxjobs * 2 + 1, || {
+            PollFd::new(-1, PollFlags::empty())
+        });
+        Ok(JobRunner {
+            snare,
+            running,
+            num_running: 0,
+            pollfds,
+        })
+    }
+
+    fn attend(&mut self) {
```

Comment:
This function probably does need a doc string. What are we attending?

Reply:
Fixed in https://github.com/softdevteam/snare/pull/7/commits/5ed09d85cc26aa0d3de30c49da1695f48513b4e0.